### PR TITLE
Add frontend http client

### LINKED ISSUE #6959: Extract common services from main plugin
## Description
Historically the main Wazuh dashboard plugin contains many common services needed for the different modules to work. We are in the process of segregating the different modules into decoupled plugins, so we need a common service provider for all of them. Therefore we need to evaluate which services are common to the modules and extract them from main plugin into a common service provider plugin.

#### Objective 
- https://github.com/wazuh/wazuh/issues/22887

## Tasks
The steps that have to be completed in order to close the issue.

- [x] Identify and make a list of the services to extract https://github.com/wazuh/wazuh-dashboard-plugins/issues/6959#issuecomment-2345873062
- [ ] Extract the services into a common provider
  - [x] https://github.com/wazuh/wazuh-dashboard-plugins/pull/7000
  - [x] https://github.com/wazuh/wazuh-dashboard-plugins/pull/7026
  - [ ] https://github.com/wazuh/wazuh-dashboard-plugins/pull/7037
- [ ] Implement the common provider in each module

## Functional requirements
- We must user the mechanisms provided in the life cycle of OpenSearch plugins to expose them, so other plugins can make use of them.

### PR DESCRIPTION
**Pull request by Desvelao**
### Description
This pull request adds a `http` client to the `wazuh-core` plugin and expose it to manage the API requests that is a composition of old services:
- WzRequest
- GenericRequest
- ApiCheck
- WzAuthentication

Moreover,
- Components
  - TableData (generic table component based on a implementation of `wazuh` plugin)
  - ServerTable (based on TableWzAPI of `wazuh` plugin)
- Hooks
  - useStateStorage: uses the localStorage and sessionStorage to persist the runtime data
 
### Issues Resolved
#6959

### Check List
- [ ] All tests pass
  - [ ] `yarn test:jest`
- [ ] New functionality includes testing.
- [ ] New functionality has been documented.
- [x] Update [CHANGELOG.md](./../CHANGELOG.md)
- [x] Commits are signed per the DCO using --signoff

### CODE REVIEW

File: plugins/wazuh-core/public/plugin.ts [hunk 1]
```diff
@@ -19,12 +20,21 @@ export class WazuhCorePlugin
...
-    const logger = {
+    // Debug logger
+    const consoleLogger = {
+      info: console.log,
+      error: console.error,
+      debug: console.debug,
+      warn: console.warn,
+    };
+    // No operation logger
+    const noopLogger = {
```
Comment (guidomodarelli) [inline plugins/wazuh-core/public/plugin.ts:31]: You can create a new file for each `Logger` inside the new folder called `logger`.

File: plugins/wazuh-core/public/services/http/server-client.ts [hunk 2]
```diff
@@ -0,0 +1,532 @@
...
+
+interface ServerAPIResponseItemsData<T> {
+  data: ServerAPIResponseItems<T>;
+  message: string;
+  error: number;
+}
+
+export interface ServerAPIResponseItemsDataHTTPClient<T> {
+  data: ServerAPIResponseItemsData<T>;
+}
```
Comment (guidomodarelli) [inline plugins/wazuh-core/public/services/http/server-client.ts:45]: You can move this interfaces to the `types.ts` file.

File: plugins/wazuh-core/public/services/http/server-client.ts [hunk 3]
```diff
@@ -0,0 +1,532 @@
...
+        typeof errorMessage === 'string' &&
+        errorMessage.includes('status code 401') &&
+        shouldRetry
+      ) {
+        try {
+          await this.auth(true); //await WzAuthentication.refresh(true);
+          return this._request(method, path, payload, { shouldRetry: false });
+        } catch (error) {
+          return ((error || {}).data || {}).message || false
+            ? Promise.reject(
```
Comment (guidomodarelli) [inline plugins/wazuh-core/public/services/http/server-client.ts:168]:
```suggestion
          return error?.data?.message || false
```

File: plugins/wazuh-core/public/services/http/server-client.ts [hunk 4]
```diff
@@ -0,0 +1,532 @@
...
+      // if(this.onErrorInterceptor){
+      //   await this.onErrorInterceptor(error, {checkCurrentApiIsUp, shouldRetry, overwriteHeaders})
+      // }
+      const errorMessage =
+        (error &&
+          error.response &&
+          error.response.data &&
+          error.response.data.message) ||
+        (error || {}).message;
+      if (
```
Comment (guidomodarelli) [inline plugins/wazuh-core/public/services/http/server-client.ts:158]:
```suggestion
      const errorMessage = error?.response?.data?.message || error?.message;
```

File: plugins/wazuh-core/public/services/http/server-client.ts [hunk 5]
```diff
@@ -0,0 +1,532 @@
...
+        shouldRetry
+      ) {
+        try {
+          await this.auth(true); //await WzAuthentication.refresh(true);
+          return this._request(method, path, payload, { shouldRetry: false });
+        } catch (error) {
+          return ((error || {}).data || {}).message || false
+            ? Promise.reject(
+                this.returnErrorInstance(error, error.data.message),
+              )
```
Comment (guidomodarelli) [inline plugins/wazuh-core/public/services/http/server-client.ts:170]:
```suggestion
                this.returnErrorInstance(error, error?.data?.message),
```

File: plugins/wazuh-core/public/services/http/server-client.ts [hunk 6]
```diff
@@ -0,0 +1,532 @@
...
+            : Promise.reject(this.returnErrorInstance(error, error.message));
+        }
```
Comment (guidomodarelli) [inline plugins/wazuh-core/public/services/http/server-client.ts:172]:
```suggestion
          return Promise.reject(
            this.returnErrorInstance(
              error,
              error?.data?.message || error.message,
            ),
          );
```

File: plugins/wazuh-core/public/services/http/server-client.ts [hunk 7]
```diff
@@ -0,0 +1,532 @@
...
+        optionsToGenericReq,
+      );
+
+      if (returnOriginalResponse) {
+        return response;
+      }
+
+      const hasFailed =
+        (((response || {}).data || {}).data || {}).total_failed_items || 0;
+
```
Comment (guidomodarelli) [inline plugins/wazuh-core/public/services/http/server-client.ts:219]:
```suggestion
      const hasFailed = response?.data?.data?.total_failed_items || 0;
```

File: plugins/wazuh-core/public/services/http/server-client.ts [hunk 8]
```diff
@@ -0,0 +1,532 @@
...
+      if (hasFailed) {
+        const error =
+          ((((response.data || {}).data || {}).failed_items || [])[0] || {})
+            .error || {};
+        const failed_ids =
+          ((((response.data || {}).data || {}).failed_items || [])[0] || {})
+            .id || {};
+        const message = (response.data || {}).message || 'Unexpected error';
+        const errorMessage = `${message} (${error.code}) - ${error.message} ${
```
Comment (guidomodarelli) [inline plugins/wazuh-core/public/services/http/server-client.ts:228]:
```suggestion
        const error = response?.data?.data?.failed_items?.[0]?.error || {};
        const failed_ids = response?.data?.data?.failed_items?.[0]?.id || {};
        const message = response?.data?.message || 'Unexpected error';
```

File: plugins/wazuh-core/public/services/http/server-client.ts [hunk 9]
```diff
@@ -0,0 +1,532 @@
...
+            : ''
+        }`;
+        return Promise.reject(this.returnErrorInstance(null, errorMessage));
+      }
+      return Promise.resolve(response);
+    } catch (error) {
+      return ((error || {}).data || {}).message || false
+        ? Promise.reject(this.returnErrorInstance(error, error.data.message))
+        : Promise.reject(this.returnErrorInstance(error, error.message));
+    }
```
Comment (guidomodarelli) [inline plugins/wazuh-core/public/services/http/server-client.ts:240]:
```suggestion
      return Promise.reject(
        this.returnErrorInstance(error, error?.data?.message || error.message),
      );
```

File: plugins/wazuh-core/public/services/http/server-client.ts [hunk 10]
```diff
@@ -0,0 +1,532 @@
...
+      }
+      const id = this.services.getServerAPI();
+      const requestData = { path, id, filters };
+      const data = await this._request('POST', '/api/csv', requestData);
+      return Promise.resolve(data);
+    } catch (error) {
+      return ((error || {}).data || {}).message || false
+        ? Promise.reject(this.returnErrorInstance(error, error.data.message))
+        : Promise.reject(this.returnErrorInstance(error, error.message));
+    }
```
Comment (guidomodarelli) [inline plugins/wazuh-core/public/services/http/server-client.ts:261]:
```suggestion
      return Promise.reject(
        this.returnErrorInstance(error, error?.data?.message || error.message),
      );
```

File: plugins/wazuh-core/public/services/http/server-client.ts (hunk 4 above)
Comment (Desvelao) [reply plugins/wazuh-core/public/services/http/server-client.ts:153]: Enhanced here 3fe48aa48dc7aea3716937c81a31cd2ee2da553d.

File: plugins/wazuh-core/public/services/http/server-client.ts (hunk 6 above)
Comment (Desvelao) [reply plugins/wazuh-core/public/services/http/server-client.ts:166]: Enhanced here 3fe48aa48dc7aea3716937c81a31cd2ee2da553d.

File: plugins/wazuh-core/public/services/http/server-client.ts (hunk 8 above)
Comment (Desvelao) [reply plugins/wazuh-core/public/services/http/server-client.ts:216]: Enhanced here 3fe48aa48dc7aea3716937c81a31cd2ee2da553d.

File: plugins/wazuh-core/public/services/http/server-client.ts (hunk 9 above)
Comment (Desvelao) [reply plugins/wazuh-core/public/services/http/server-client.ts:229]: Enhanced here 3fe48aa48dc7aea3716937c81a31cd2ee2da553d.

File: plugins/wazuh-core/public/services/http/server-client.ts (hunk 7 above)
Comment (Desvelao) [reply plugins/wazuh-core/public/services/http/server-client.ts:211]: Enhanced here 3fe48aa48dc7aea3716937c81a31cd2ee2da553d.

File: plugins/wazuh-core/public/plugin.ts (hunk 1 above)
Comment (Desvelao) [reply plugins/wazuh-core/public/plugin.ts:33]: We will need to discuss about the frontend logger. For now, I will leave the code in that state.

File: plugins/wazuh-core/public/utils/configuration-store.ts [hunk 11]
```diff
@@ -1,77 +1,101 @@
...
   async clear(...settings: string[]): Promise<any> {
     try {
       const attributes = await this.get();
       const updatedSettings = {
         ...attributes,
       };
-      settings.forEach(setting => delete updatedSettings[setting]);
+
+      for (const setting of settings) {
+        // eslint-disable-next-line @typescript-eslint/no-dynamic-delete
```
Comment (guidomodarelli) [inline plugins/wazuh-core/public/utils/configuration-store.ts:93]: Could you disable this rule

File: plugins/wazuh-core/public/types.ts [hunk 12]
```diff
@@ -1,20 +1,54 @@
...
+    ) => React.ComponentType<TableDataProps<T>>;
+    // eslint-disable-next-line @typescript-eslint/naming-convention
+    SearchBar: (prop: any) => React.ComponentType<any>;
+    // eslint-disable-next-line @typescript-eslint/naming-convention
+    ServerTable: <T>(
+      prop: ServerDataProps<T>,
+    ) => React.ComponentType<ServerDataProps<T>>;
+  };
 }
 // eslint-disable-next-line @typescript-eslint/no-empty-interface
```
Comment (guidomodarelli) [inline plugins/wazuh-core/public/types.ts:29]: You can remove this line

Comment (guidomodarelli) [reply plugins/wazuh-core/public/types.ts:29]:
```suggestion
```

File: plugins/wazuh-core/public/services/http/server-client.ts [hunk 13]
```diff
@@ -0,0 +1,513 @@
...
+    // eslint-disable-next-line unicorn/no-object-as-default-parameter
+    extraOptions: {
+      shouldRetry?: boolean;
+      checkCurrentApiIsUp?: boolean;
+      overwriteHeaders?: any;
+    } = {
+      shouldRetry: true,
+      checkCurrentApiIsUp: true,
+      overwriteHeaders: {},
+    },
```
Comment (guidomodarelli) [inline plugins/wazuh-core/public/services/http/server-client.ts:68]: Default parameters should not be passed to a function through an object literal. The `foo = {a: false}` parameter works fine if only used with one option. As soon as additional options are added, you risk replacing the whole `foo = {a: false, b: true}` object when passing only one option: `{a: true}`. For this reason, object destructuring should be used instead.

File: plugins/wazuh-core/public/services/http/server-client.ts [hunk 14]
```diff
@@ -0,0 +1,513 @@
...
+      logged: false,
+      token: null,
+      account: null,
+      policies: null,
+    };
+    this.userData$ = new BehaviorSubject(this.userData);
+  }
+
+  /**
+   * Permorn a generic request
```
Comment (guidomodarelli) [inline plugins/wazuh-core/public/services/http/server-client.ts:50]:
```suggestion
   * Perform a generic request
```

File: plugins/wazuh-core/public/services/http/server-client.ts [hunk 15]
```diff
@@ -0,0 +1,513 @@
...
+   */
+  async request(
+    method: HTTPVerb,
+    path: string,
+    body: any,
+    // eslint-disable-next-line unicorn/no-object-as-default-parameter
+    options: {
+      checkCurrentApiIsUp?: boolean;
+      returnOriginalResponse?: boolean;
+    } = { checkCurrentApiIsUp: true, returnOriginalResponse: false },
```
Comment (guidomodarelli) [inline plugins/wazuh-core/public/services/http/server-client.ts:178]: Default parameters should not be passed to a function through an object literal. The `foo = {a: false}` parameter works fine if only used with one option. As soon as additional options are added, you risk replacing the whole `foo = {a: false, b: true}` object when passing only one option: `{a: true}`. For this reason, object destructuring should be used instead.

Comment (guidomodarelli) [reply plugins/wazuh-core/public/services/http/server-client.ts:178]: https://github.com/sindresorhus/eslint-plugin-unicorn/blob/main/docs/rules/no-object-as-default-parameter.md

File: plugins/wazuh-core/public/services/http/server-client.ts (hunk 13 above)
Comment (guidomodarelli) [reply plugins/wazuh-core/public/services/http/server-client.ts:68]: https://github.com/sindresorhus/eslint-plugin-unicorn/blob/main/docs/rules/no-object-as-default-parameter.md

File: plugins/wazuh-core/public/services/http/server-client.test.ts [hunk 16]
```diff
@@ -0,0 +1,106 @@
...
+          policies: {},
+          account: null,
+          logged: true,
+        });
+        done();
+      });
+    });
+  });
+
+  it.only('Unauthentication', done => {
```
Comment (guidomodarelli) [inline plugins/wazuh-core/public/services/http/server-client.test.ts:90]:
```suggestion
  it('Unauthentication', done => {
```

File: plugins/wazuh-core/public/utils/configuration-store.ts (hunk 11 above)
Comment (Desvelao) [reply plugins/wazuh-core/public/utils/configuration-store.ts:93]: what do you mean?

File: plugins/wazuh-core/public/types.ts (hunk 12 above)
Comment (Desvelao) [reply plugins/wazuh-core/public/types.ts:29]: Fixed.

File: plugins/wazuh-core/public/services/http/server-client.ts (hunk 13 above)
Comment (Desvelao) [reply plugins/wazuh-core/public/services/http/server-client.ts]: Fixed

File: plugins/wazuh-core/public/services/http/server-client.ts (hunk 15 above)
Comment (Desvelao) [reply plugins/wazuh-core/public/services/http/server-client.ts]: Fixed

File: plugins/wazuh-core/public/services/http/server-client.test.ts (hunk 16 above)
Comment (Desvelao) [reply plugins/wazuh-core/public/services/http/server-client.test.ts]: Fixed.

**Review by guidomodarelli — APPROVED** [review]

**Review by chantal-kelm — APPROVED** [review]

**Pull request merged**